### ISSUE TITLE
Bootstrap 4 and gpRow

### ISSUE THREAD

**Issue by mahotilo** [Typesetter/Typesetter#244]
With BS3 we have no problem to combine `gpRow` class with manually applied `col-x-y` classes.
Now with BS4 `gpRow` is not equal to `row` class. Therefore` grRow` section wrapper cant properly control sections with manually applied `col-x` classes. 
We need to make `gpRow` equal to `row` again.

**Comment by juek** [comment]
I disagree. Grid systems should never be mixed. 

Typesetter grid has 12 columns, only 2 breakpoints and a fixed gutter width and associated row margins, 

Bootstrap 3 grid has configurable column count ($grid-columns), gutter width ($grid-gutter-width). Its .row has negative half $grid-gutter-with side margins.

Bootstrap 4 grid is non-foating and flex based.

We cannot simply mix or unify the concepts.

But I believe that Typesetter should be aware of the framework used by a theme and use appropriate class names for Section Combos. It should also know the available breakpoints and menu collapsing treshold ($grid-float-breakpoint). I just yet have no idea how to implement such an interfce.

**Comment by mahotilo** [comment]
Maybe we can select one of `additional.css` with different `gpXxx` classes definition when we process call of `\gp\tool::LoadComponents( 'bootstrapX-XX' );`?

**Comment by juek** [comment]
IMO the best solution would be to automatically add, let's say, 'bootstrap-3' and 'bootstrap-4' class name to the body element, together with some additional BS-specific metadata. Then we could do everything with CSS.

**Comment by juek** [comment]
Now with the ['bootstrap-_n_' classes](https://github.com/Typesetter/Typesetter/blob/master/themes/Bootswatch4_Scss/template.php#L21) in the Bootswatch Scss themes, we should be able to do sth. like 

```
.gpRow {  @extend .row; }
.gpCol-1 {  @extend .col-sm-1; }
.gpCol-2 {  @extend .col-sm-2; }
.gpCol-3 {  @extend .col-sm-3; }
…
```

(not yet tested) but this will not take custom [$grid-columns](https://getbootstrap.com/docs/4.0/layout/grid/#variables) values into account and only work with the default 12 columns. So I'm reluctant to add it by default.

**Comment by mahotilo** [comment]
I confirm it, if add
```
.gpRow {  @extend .row; }
.gpCol-1 {  @extend .col-sm-1; }
.gpCol-2 {  @extend .col-sm-2; }
```

to template styles.scss it work fine.

It is also possible to load Bootstrap preset in "Manage classes" and use it.

>...this will not take custom $grid-columns values into account and only work with the default 12 columns.

As we can read in additional.css
> gpEasy-SimpleGrid
> 12-columns, percent based widths, px-based paddings.

So it also ignore $grid-columns.

I worried only about gpRow from Default preset. If we should change something, lets change gpRow to force it work fine both with gpCol-y, BS3 col-x-y and BS4 col-x-y.

By the way, should we make both BS3 preset and BS4 preset? They will have differences.

Or maybe I just misunderstood your plan.

**Comment by mahotilo** [comment]
Just noticed
![image](https://user-images.githubusercontent.com/14929385/41794101-ebdb5aee-7666-11e8-8da7-cd7587f7e73f.png)
Is it only at me?

**Comment by juek** [comment]
> Is it only at me?

No, [fixed](https://github.com/Typesetter/Typesetter/commit/f61396c7b8508dbefd03e7b5b7aea46128430319) (en.main.inc wasn't UTF8-encoded)

**Comment by juek** [comment]
By the way, should we make both BS3 preset and BS4 preset? 

Yep, we probably should.

**Comment by juek** [comment]
> So it also ignore $grid-columns.

Typesetter's native grid doesn't support such and we can't simply change it because of the default section combos (text-image and text-gallery) which rely on a 12-col grid.

But we should probably add

```
.gpRow { 
  @extend .row; 
}

[class*="gpCol-"] { 
  padding-left: $grid-gutter-width / 2; 
  padding-right: $grid-gutter-width / 2; 
}
```

to all Bootswatch themes in order to correclty adapt the grid.

**Comment by mahotilo** [comment]
We have non-BS themes, B3-themes and now BS4-themes.
gpXXX system basically defined only for non-BS themes.
In case of BS3,4 it can be simple alias for row and col-xs-x classes?
If I am right, can we use new `class="bootstrap-X"` as marker for defining gpXXX system as alias. If no `class="bootstrap-x` are defined, we should use old style gpXXX system.

Yes, I again try to draw your attention to this option:)

**Comment by juek** [comment]
> Or we need to introduce in PHP global variable $BootstrapVersion 

I was thinking sth. in Addon.ini

[Framework]
name = Bootstrap
version = 4.1.1

**Comment by juek** [comment]
... a gobal var in template.php would be set too late to make use of it in earlier processed components/hooks

**Comment by juek** [comment]
... which then would be available in $gpLayouts …

```
global $page, $gpLayouts;

$layout_id = $page->gpLayout;
$current_layout = $gpLayouts[$layout_id];

$frontend_framework = !empty($layout['framework']) ? $layout['framework'] : false;

if( $frontend_framework ){
  msg(
    'Current page uses the front-end framwork '
    . $layout_framework['name'] 
    . ' version ' . $layout_framework['version']
  );
}
```

**Comment by mahotilo** [comment]
Do we need change Layout.php -> IniExtract() to see new section [FrontEndFramework] or it is processed in another place?

**Comment by mahotilo** [comment]
I pulled 5 request  (sorry, I forget all the time how to pull many files in one request). It is rough solution for default and Bootstrap4 mode. It works for me as expected.

It is so strange. I can't use file name `additional_bs4.scss` it doesn't work, but `additional_bs41.scss` is ok.

**Comment by juek** [comment]
Thanks. I'll use it as a proof-of-concept but will change a few things. 
E.g. I firmly believe we should do the grid adaption in the theme's scss. 
additional.css is sort of a thorn in my side - I'd rather like to get rid of it at all.

**Comment by mahotilo** [comment]
If look attentively at the console one may see errors in image_edit.js during editing new inserted image section. (In the current beta and with my addings)

For example.
```
function SaveData(){

			save_obj.posx		= field_x.value; <---
```
Uncaught TypeError: Cannot read property 'value' of null

I can't find way to fix it.

**Comment by juek** [comment]
On the subject of presets: I believe we should go for possible multiple named presets in parallel, we then could choose from, maybe even inside the Section Attributes dialog. My inital idea was to make it possible to define such presets in the course of a theme, which wasn't possible when I created the feature as a plugin back then. 

Maybe it should be reconsidered.

Let's simply think of adding a classes.php array (like /data/_config/classes.php once saved) to the theme folder. Or rather the layout subfolder? - no. Should be possible to automatically enable the appropriate set based on the used theme/layout in editing context

**Comment by mahotilo** [comment]
Hmm, I begin to see the way you are proposing.
You want to make single place configure not only gpXXX classes but all custom theme classes, and make they reachable not only in Section attributes but in CK-editor also (like in your CKE Styles Combo).

But. 
CKE Styles Combo automate creation of presets based on classes names. It's cool because of dynamic character of theme style developing process and theme life management. 
But, gpXXX presets are more static, and now we have only 3 of they: non-BS, BS3 and BS4.
Do we need to bind this in one?

**Comment by juek** [comment]
> like in your CKE Styles Combo

No, I actually didn't, but now as you're saying it… :)

**Comment by juek** [comment]
> [FrontEndFramework]
> name = Bootstrap
> version = 4.1.1

Implemented with [this commit](https://github.com/Typesetter/Typesetter/commit/e5b8bc0cde253df70c860ef0e8b4321921d84c41)

**Comment by juek** [comment]
> .gpRow {  @extend .row; }
> .gpCol-1 {  @extend .col-sm-1; }
> .gpCol-2 {  @extend .col-sm-2; }

should be implemented in themes IMO (have already done so with unpublished themes and it works nicely)
